cli/src/semgrep/core_runner.py

```diff
@@ -634,9 +634,11 @@ def plan_core_run(
 
         Note: this is a list because a target can appear twice (e.g. Java + Generic)
         """
-        target_info: Dict[Tuple[Path, Language], List[int]] = collections.defaultdict(
-            list
-        )
+        # The range of target_info is (index into rules x product as json)
+        # Using product as JSON because we want structural equality of products instead of object equality.
```

Comment:
but 'product' in the ATD is defined as dataclass(frozen=True), so you would get structural equality there too.